### DISCUSSION TITLE
A new `defineIntegration` helper

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#886] [Proposal]
### Body

# Summary

Adds a new `defineIntegration` helper to Astro to help integrations authors define integrations in a convenient/robust way.

# Background & Motivation

We've been working on the [Astro Integration Kit](https://astro-integration-kit.netlify.app/) since January with other fellow maintainers and Astro community members. The goal is to improve the DX of integrations authors and make their lives easier, allowing more people to start writing integrations as well.

It's a set of utilities we plan to contribute back to Astro and we're starting today with `defineIntegration`!

Defining an integration but the DX could be improved using a function helper rather than just having the `AstroIntegration` type. Also, handling correctly options (so that it's both type and runtime safe) is overlooked by many people. We took inspiration from Starlight and experimented with it for a few months before feeling we have it right, but we need your feedback!

# Goals

- Have a function acting as a type helper, similarly to `defineConfig`
- Allow to type options easily
- Allow (optionally) to have options runtime safety
- Make it backward compatible

# Example

> Example implementation at https://github.com/withastro/astro/pull/10892 (outdated with this proposal but fairly close)

> I originally suggested the zod version only and edited the following examples with alternative options

## `defineIntegration`, the vanilla way

`defineIntegration` would not be a simple type helper like `defineConfig` that returns `AstroUserConfig`, but pretty close. Instead, it would return `(options: Options) => AstroIntegration`.

```ts
const myIntegration = defineIntegration({
	name: "my-integration",
	// options is never in this case
	setup({ name, options }) {
		return {
			hooks: {}
		}
	}
})

myIntegration()
```

```ts
type Options = {
	foo: string
}

const myIntegration = defineIntegration({
	name: "my-integration",
	setup<Options>({ name, options }) {
		return {
			hooks: {}
		}
	}
})

myIntegration({ foo: "bar" })
```

> Note that `setup` is named after Nuxt Kit `defineModule`, which is inspired by the vue ecosystem. Happy to rename it.

`defineIntegration` is already available for usage in AIK, [read docs](https://astro-integration-kit.netlify.app/core/define-integration/). [View source code here](https://github.com/florian-lefebvre/astro-integration-kit/blob/main/package/src/core/define-integration.ts).

## `defineIntegration`, the zod way

Matthew had concerns about only providing a zod compatible utility so I introduced the above alternative. However, people wanting to use zod could use a `defineIntergation.withZod` utility that would behave exactly the same as the AIK version. See [docs](https://astro-integration-kit.netlify.app/core/define-integration/) and [source](https://github.com/florian-lefebvre/astro-integration-kit/blob/main/package/src/core/define-integration.ts).

## Backward compatibility

`defineIntegration` returns `AstroIntegration` so it's compatible. We talked about a potential `@astrojs/kit` package to have backward compatible utilities, but after some discussions with Erika I think it's better not to have it. If integrations authors want to benefit from new hooks params, they should update their peer deps. That should encourage more users to upgrade to more recent versions of Astro, and the process is painless anyway. Note that this means `defineIntegration` will only be available from the minor it's part of.

**Comment by Princesseuh** [comment]
+1 😄

I like the `@astrojs/kit` idea with the re-export. It's a bit wonky for the utilities added on hooks though, I wonder if there's a good way to do this. My own personal suggestion would be to not care too much, as users who are likely to use newer / updated integrations tend to reach updated versions of Astro pretty quickly. Looking at telemetry, 50% of users are not on 4.5 right now.

**Reply by ematipico** [reply]
I second this idea. The use of `@astrojs/kit` seems to be a very good comprise!

**Comment by FredKSchott** [comment]
Love the idea of an official helper! Big +1! 

Should this live in Astro though? If I'm designing an integration as an external package, then Astro isn't usually a dependency of my package (I don't think). As an integration author I'd actually prefer it to be an external dependency of my package, and then I could manage major versions myself vs. having to build for the `astro@3` interface, the `astro@4` interface, the `astro@5` interface, etc. etc.

I'd expect this use-case of a `defineIntegration` helper built-in to Astro to only really be useful for people building integrations in their own app codebase vs. npm packages.

**Reply by florian-lefebvre** [reply]
In this case, Astro is a peer dependency. So if you want to support only from when it's added, then you add the `"astro": "^4.7.0"` peer dep and you're good to go. However, it's true to if you want to support older versions then you need `@astrojs/kit` as a dependency.

To be clear, I feel like `defineIntegration` should actually be more useful for integrations authors (ie. npm packages).

Idk if that answers your question at all 😅, feel free to re-explain if I didn't get it!

**Reply by Princesseuh** [reply]
`@astrojs/kit` would anyway need to depend on Astro for types and stuff, unless we move *all* of it into a separate package. It always makes sense for integrations to have Astro as a peerDep.

**Comment by matthewp** [comment]
Thanks for kicking this off @florian-lefebvre! My feedback at this stage is that the goals are a little vague and could use being more specific. A breakdown of what I mean:

* Goal: Improve the DX of defining integrations
  * I think this is mostly to get type inference via the `defineIntegration` function? Are there other DX improvements that this proposal will provide? If so list them specifically.
* Make it easy to handle options, at both type and runtime levels
  * I don't think I understand this one. Do you mean to handle options that are passed to your integration? If the goal is to make it easier, what is hard about it at present. `{ arg }: MyOptions` seems perfectly easy to me but I'm very likely missing something.
* Promote good practices around the ecosystem
  * What are those good practices that we want to promote?
* Make it backward compatible
  * Do you mean that this proposal intends to add sugar to the existing API without replacing it?

**Reply by florian-lefebvre** [reply]
> I think this is mostly to get type inference via the defineIntegration function? Are there other DX improvements that this proposal will provide? If so list them specifically.

Yes type inference

> I don't think I understand this one. Do you mean to handle options that are passed to your integration? If the goal is to make it easier, what is hard about it at present. { arg }: MyOptions seems perfectly easy to me but I'm very likely missing something.
> What are those good practices that we want to promote?

Those are actually linked. I didn't want to say that most integrations are not following best practices btw! But what I mean is that setting a type for an integrations options doesn't ensure the value received is actually of that type. TLDR I think it's great to use zod for this. It also allows great pattern with ease (eg. https://github.com/astrolicious/i18n/blob/feat/sitemap/package/src/options.ts#L155-L161)

> Do you mean that this proposal intends to add sugar to the existing API without replacing it?

I mean that `defineIntegration` still returns `() => AstroIntegration`, so no breaking changes and can be used in older versions of Astro

**Reply by matthewp** [reply]
Ok, if the goal is to provide validation to integration options then I would say that specifically.

----

I'm not sure how I feel about that goal, to be honest. I think it's possible to overdo using Zod and other validation libraries and this one is close to being that. Not every function in your application should have its arguments validated. There is a performance tradeoff at play here.

**Reply by florian-lefebvre** [reply]
Yeah I get your point for sure. That being said, idk if performance is really a big issue here since validation only runs when the integration is added to astro (either in a user's config or by another integration), so it should not be called that much, and not at runtime.

**Comment by florian-lefebvre** [comment]
I updated the RFC, please have another look!

**Comment by matthewp** [comment]
I still don't think we should do schemas as part of this API. Having the type param as shown in the updated example I think is what we should recommend. I would rather schemas be something that stays inside of a 3rd party library like integration kit. At least for now, we can add layer.

Keeping this API small and focused on providing types for the shape of the integration hooks (as shown in the example) is the way to go, imo.

**Reply by Adammatthiesen** [reply]
Just my 2 cents from a integration Author, please take nothing i say as being against you as its just my opinion on this utility.

Personally i think being able to use the `astro/zod` OptionsSchema field has been one of the best things about AIK, And personally i would love to be able to move some of my integrations back to core such as the new StudioCMS project but i feel having to redefine the schema into a type would partially defeat the purpose of my usage of zod in the first place requiring me to validate my users options instead of having it included within the integration API,

When i started building integrations around 4 months ago, this would have also saved me hours of trying to understand how integrations do user options validation instead of being able to spend time actually building the main integration.  I would 100% love to see `defineIntegration` come into astro in its best form instead of a diminished form that would just keep me and my integrations using AIK instead of astro Core

**Reply by Fryuni** [reply]
I think having option validation from Astro enables best practices to be more widely used in the community. Using Zod for this to me seems like the best fit.

Zod is already heavily used and deeply integrated within Astro:
- It is used in Core to validate options and collections
- It is used in official integrations, like sitemap
- It is used in Starlight

Also, when using `create-astro` we generate `.mjs` config files. Those don't have any typing protection. You get autocomplete and intellisense, but if you forget to pass a parameter, pass a value with an incorrect type, or pass something that doesn't exist, you get zero visibility as a user:
![image](https://github.com/withastro/roadmap/assets/11063910/ffdec05c-2e43-4538-afa0-7401c4e2ae2c)

Compare that to using Starlight that has proper validation:
![image](https://github.com/withastro/roadmap/assets/11063910/02c92879-0b4a-4e74-8eab-87f463ef9d02)

**Reply by Fryuni** [reply]
In fact, Zod is used to validate the integrations themselves passed in the config. So using them to validate the options passed _to_ the integrations seems like a natural continuity

**Reply by Fryuni** [reply]
Oh, and just to be clear. The LSP also doesn't report anything:
![image](https://github.com/withastro/roadmap/assets/11063910/1658e744-9c43-4fdd-95fa-7957045f7903)

It would report if we generated `.mts`:

![image](https://github.com/withastro/roadmap/assets/11063910/fd053281-b210-4bac-ab1a-5f92eeff4169)

But that doesn't apply to any existing project that is using `.mjs`

**Reply by matthewp** [reply]
I have to admit there are some good points here. My hesitancy comes from:

1. There is run-time cost to looping over objects and comparing types, etc. You don't want every function your program doing this. So where do we draw the line?
2. The integrations we have in core have a few options and doing some manual checking or coercing works perfectly fine. There aren't so many options that needing to check each one is cumbersome.
3. There are `options`, which mean they should be *optional* most of the time. In core we only have 1 option that's required (IIRC), `@astrojs/node` requires `mode`. And if options are optional then you can assign the default when it's not provided.

**Reply by Fryuni** [reply]
> There is run-time cost to looping over objects and comparing types, etc. You don't want every function your program doing this. So where do we draw the line?

I'd say it is fine to validate configuration options. Functions called after that, and especially functions called at runtime should rely on TS and not runtime validations.

And I draw the line there just for compatibility with the `.mjs` config file.

As for performance, Node's startup time takes longer than validating the Starlight config, which is on the more complex end of the integration options spectrum. Since those run once for the config I don't think it is a big problem. Sure, if you add dozens of integrations you'll get a couple of extra seconds to run any Astro command, but that seems ok for me. The wiring of integrations inside Astro will most likely take much longer than validating the options for those integrations since it goes back into the event queue many times.

> The integrations we have in core have a few options and doing some manual checking or coercing works perfectly fine. There aren't so many options that needing to check each one is cumbersome.

Similarly, those simple cases are trivial in Zod. We are not forcing authors to use Zod; we just provide it as a first option. You can still not use it. The proposal is intended to be backward compatible.

> There are options, which mean they should be optional most of the time. In core we only have 1 option that's required (IIRC), @astrojs/node requires mode. And if options are optional then you can assign the default when it's not provided.

I don't see how that is an issue here...
You can do `.default({ ... })` or `.optional()` and `setup({options = {...}})` to have a default value.

**Reply by delucis** [reply]
So one thing I’m thinking about is how supporting a schema is forcing `defineIntegration()` to return a function instead of leaving that to users. This also makes supporting option custom typing/parsing more difficult. I’m wondering if that also means it might get in the way of other future things users want to do (or at least add mental overhead).

Currently, if an integration author wants options, they do something like:

```ts
export default (options) => {
  // use options then return
  return { /* ... */ } satisfies AstroIntegration;
};
```

The `defineIntegration()` form effectively blocks that — now you need to do everything in the `setup()` hook instead and typing options needs to be done by a generic type param:

```ts
import { z } from "astro/zod"

export default defineIntegration({
    name: "my-integration",
    optionsSchema: z.object({
        virtualModuleId: z.string()
    }),
    setup({ options, name }) {
        return {
            hooks: { /* ... */ }
        }
    }
})
```

My preference would honestly be for something much more transparent and closer to a vanilla type helper:

```ts
import { z } from "astro/zod"

const optionsSchema = z.object({
    virtualModuleId: z.string()
});

export default (userOptions: z.input<typeof optionsSchema> = {}) => {
  const options = optionsSchema.parse(userOptions);
  return defineIntegration({
    name: "my-integration",
    hooks: { /* ... */ }
  });
};
```

This leaves the API flexible to any options validation approach an integration author prefers and is barely any more code — the distinction is that in the Zod case, you need to parse the options yourself.

I’ll admit that one of the nice things about handling this ourselves would be taking care of errors are thrown and rendered etc., but if that’s a concern a helper function we export from `astro/zod` could also be an option.

**Reply by HiDeoo** [reply]
I'm also personally leaning in favor of letting integration authors handle the validation themselves, if they want/need to, with the tooling they want:

- Very personal, but I have some old integrations doing some validations with a custom internal library that is not Zod.
- As visible in the last example shared by Chris, it's still very easy to use Zod if you want to and doesn't add a lot of code.
- If you get access to `options` parsed/validated by Zod, how do you get access to raw options provided by the user? For example, in Starlight, the raw options are used by the plugin system as a plugin could read or update user defined options which may have a different shape/content than the parsed one if the schema include transformations or defaults. I guess they could also be provided as a new property, but isn't that a bit redundant vs just letting the integration author handle the validation however they need?

**Reply by florian-lefebvre** [reply]
Okay I guess AIK's version is still too opiniated. At least I'd like a little zod helper that deals with most cases:
```ts
const optionsSchema = z.object({
	foo: z.string()
})

export function integration(rawOptions: z.input<typeof optionsSchema>) {
	// will pass the zod error map and throw an astro error if invalid, could be exported from "astro/integration" as well
	const options = validateIntegrationOptions(optionsSchema, rawOptions)

	return defineIntegration({
		name,
		hooks: {}
	})
}
```

**Reply by delucis** [reply]
Yeah, I think a little utility like this could be nice — in Starlight we have our own error map etc. but that’s a quite advanced thing to maintain (IIRC ours is even a bit more advanced the default Astro one which had some gaps in what it covered well)

https://github.com/withastro/starlight/blob/0e95895330d3a4181d7a0069d815ca3bf950ab78/packages/starlight/utils/error-map.ts#L15-L33

**Reply by florian-lefebvre** [reply]
I think it would be worth upstreaming the changes to core. I looked into it and it could be even reused in astro db

**Reply by bluwy** [reply]
My opinion among others: I also agree with Matthew. I think when it comes to validation (e.g. zod, etc), it's most useful when dealing unknown inputs and trying to parse/coerce it to a data type you want. For example, forms, content collection frontmatter, soon-to-be `envSchema`, etc.

If we already know the incoming input shape, especially one already validated by TypeScript, then the usecase for validation becomes slimmer, basically covering what TypeScript couldn't cover. And for integrations, TypeScript is usually the only defence you need, other validations (usually not many) can be checked manually.

The Astro config is definitely an outlier here. If I'd rewrite it today, I'd probably avoid using zod altogether, which will also force us to write stricter types, not to mention easier to maintain our types too. I think this is where I draw the line for validation (not only zod).

**Reply by ematipico** [reply]
I am also leaning towards **not forcing** developers to have a schema for their options. Let's shape the APIs in a way where they create options validated with a schema they want.

**Reply by Fryuni** [reply]
> especially one already validated by TypeScript

@bluwy Just going back to this, Astro configurations are _not_ validated by TS unless the user of the integration (the final project) renames their config file. And they have no indication that they should do that.

It might have been collapsed by GH, but I sent it here: https://github.com/withastro/roadmap/discussions/886#discussioncomment-9330867

It differs from the other `define*` functions because those are used in TS contexts, so we can rely on TS validations.

**Reply by bluwy** [reply]
My point is that if I (as a package author) write `function foo(options?: IntegrationOptions) {}`, I'd expect users to pass in an `options` object that will satisfy TypeScript and hence "validated". If they don't enable typechecking via `.ts` or `checkJs`, then it's on them.

**Reply by florian-lefebvre** [reply]
I disagree just like Fry on this point, because of the Astro defaults

**Reply by Fryuni** [reply]
Yeah, `create-astro` and all the templates have TS validation disabled by default on the config. The config file is a `.mjs` and the base tsconfigs have `checkJs: false`.

The main point of `defineIntegration` is, IMO, exactly to steer people into validating their options because the type will not be validated by TS unless the user explicitly opts into it.

**Reply by Fryuni** [reply]
`defineConfig` and others are useful as no-op type helpers because those are used on default exports or exports where we want to autocomplete something within an inferred type (collections).
In those cases, the user can't write the type anywhere to get IntelliSense, so a type helper is there to save you.
But for integrations, you can just write `: AstroIntegration` in your function.

If this is the boilerplate we'd show people:
```ts
import { defineIntegration } from 'astro';
import { z } from "astro/zod"

const optionsSchema = z.object({
    virtualModuleId: z.string()
});

export default (userOptions: z.input<typeof optionsSchema> = {}) => {
  const options = optionsSchema.parse(userOptions);
  return defineIntegration({
    name: "my-integration",
    hooks: { /* ... */ }
  });
};
```

I don't think it has any advantage over this:
```ts
import type { AstroIntegration } from 'astro';
import { z } from "astro/zod"

const optionsSchema = z.object({
    virtualModuleId: z.string()
});

export default (userOptions: z.input<typeof optionsSchema> = {}): AstroIntegration => {
  const options = optionsSchema.parse(userOptions);
  return {
    name: "my-integration",
    hooks: { /* ... */ }
  };
};
```

But this is much cleaner, and by requiring validation ensures authors don't miss the fact that types won't help here:
```ts
import { defineIntegration } from 'astro';
import { z } from "astro/zod"

const optionsSchema = z.object({
    virtualModuleId: z.string()
});

export default defineIntegration({
  name: "my-integration",
  optionsSchema,
  setup: ({options}) => ({
    hooks: { /* ... */ }
  }),
});
```

**Reply by delucis** [reply]
I think “much cleaner” might be overstating it slightly 😁 — in the grand scheme of things once there’s actual code in these snippets, the difference is kind of small. As @HiDeoo mentioned above, plugins like Starlight that already do validation with Zod couldn’t even use the proposed `defineIntegration()`, because we need access to both raw and validated user config. Of course, it’s _possible_ to add access to that in `setup()`, but we’re obfuscating stuff that is just regular JS by doing it like that, which is not a great sign.

My instinct is:
- a new API should support functionality that is significantly harder to do without the API
- a new API should be flexible enough to cater for expected needs

I’m not sure if the current design meets those kinds of thresholds.

There are certainly arguments for making nice Zod parse errors easier, but that can be needed all over an integration code base — for example Starlight throws pretty mapped errors in several places — so a generic `parseWithfriendlyError()` utility could go further than baking it into `defineIntegration()`.

**Reply by bluwy** [reply]
What if we set `// @ts-check` for the config by default, or work towards setting `checkJs: true` by default? Do we still need the zod validation then? If the JS ecosystem go by the rule where every input couldn't be trusted and needs validation, the ecosystem will be very slow.

I get that integrations have a specific target audience, and lives in the Astro config, but you couldn't expect everything the user imports to have runtime validation too. They need to trust TypeScript in some APIs and we could extend that for everything else including integrations.

**Comment by bluwy** [comment]
TLDR: -1 for me, but I'll explain below 😄 

### About `define*`

Stepping back, I think we need to figure out the reason `define*` API exists today, including `defineConfig`, `defineCollection`, `defineMiddleware`, etc:
- Easy way to define something with types
- Syntax is the same in JS & TS
- Easy to remember, e.g. type `define|` and get it autocompleted/imported

Otherwise it's much a shell function that doesn't have a lot of utility. It's most useful for end-users, but not much for package authors? So speaking for `defineIntegration` and `defineToolbarApp` (which maybe I should've voiced before), the difference between `export default define*()` and `export default {} satisfies X` is not a lot. Not to mention, you'd often export a function instead, so it becomes `export default () => define*()` and `export default (): X => {}`, which the latter is more natural to me. So some `define*` API may not make sense.

(I know `satisifes X` and `: X` is not exactly the same thing, but I think it's not an issue in practice)

### About `defineIntegration`

I notice `defineIntegration` is different and it's not a shell function (returns param directly), which conflicts with how the other `define*` API work. It introduces two syntaxes to author integrations, which I'm not a fan of.

On top of that, I agree with [Matthew's comment](https://github.com/withastro/roadmap/discussions/886#discussioncomment-9330394), and if we remove the schema handling, there's not much left for a sort of `defineIntegration` API.

**Reply by Princesseuh** [reply]
I would argue that package authors are also end-users and those benefits also applies to them. While we can most definitely consider them more advanced users, they definitely deserve the same level of convenience when possible imo! In my opinion, the discoverability aspect is huge notably.

**Reply by bluwy** [reply]
The convenience also isn't free, if you start importing runtime from `astro`, you'd need to think about _"Should `astro` be in dependencies, devDependencies, or peerDependencies?"_ and it's very easy mess it up especially if the target audience is for those new to developing integrations or toolbar apps.

**Reply by Princesseuh** [reply]
I don't know, it seems fairly intuitive to me that an integration would have Astro as a peerDependency, while it's true that this burden exists with runtime import, I also feel like it's fairly minor! (especially if we lead this effort into creating a starter project etc in the future)

**Reply by delucis** [reply]
On the whole I also agree that we have a convention that our `define*` helpers are all identity functions that are only used as type helpers (notably in contexts like `astro.config.mjs` where a user might not be writing TypeScript), so breaking that convention seems something worth discussing, especially if we break it for little added functionality.

**Reply by bholmesdev** [reply]
@delucis This is a good point on what `define*` utilities should do. Until now, they were sugar over `satisfies` (or type casts before `satisfies`) that work in JS and TS files. Now, `defineIntegration()` is doing a bit of internal work to potentially parse Zod schemas. `defineAction()` has a similar issue.

This may be an odd take... but I don't think it matters what the internals of `define*` actually do. Definitions expand overtime, and I think `define` has expanded to mean "how you configure a feature in Astro." That feature may be a collection, a database table, an action, or yes, an integration. We may think advocating for manual TS types or `create*` instead of `define*` add more clarity. In a way, it actually adds confusion. As a user, you would need to know the internals of a function to discern whether `define` or `create` is the import to expect. Users should never have to consider that distinction. I think `defineIntegration` is the right call to build a configuration helper that we can build upon for future features.

**Reply by ematipico** [reply]
I somewhat agree, but I'm afraid I also have to disagree. I agree that  users should not care about the internals of a function, but Astro is a framework that allows to define third-party plugins, which means that we build an ecosystem. 

Eventually, it will be normal for developers to look up to Astro for naming conventions, coding patterns, etc., so the end-users will have a consistent DX.

For example, look at the [naming convention set by Rust for function names](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv). I believe they are really good, because via convention, you know **what do expect** from the naming of function.

Hence, I believe we should be strict here, and come up with a different prefix identity functions that do **more** than just returning the defined object.  


E.g. 
- `define*` -> cheap operation
- `create*` -> expensive operation, for example internal parsing and transformations

**Reply by florian-lefebvre** [reply]
Interesting, that means `defineAction` would have to be renamed

**Comment by florian-lefebvre** [comment]
I want to get back to this RFC now that I'm mostly done with `astro:env`. I think an important aspect of `defineIntegration` is that it much be easy to use for beginners. Nowadays, I may not need it because I'm familiar with the API but it seemed a bit scary at first. I don't know if `defineIntegration` is the solution to this but I think it can help

**Comment by florian-lefebvre** [comment]
Not relevant anymore, the existing API is great. Closing
